oak_functions/loader/src/metrics.rs

```diff
@@ -0,0 +1,283 @@
+        Ok(Self {
+            count: 0,
+            epsilon: config.epsilon,
+            batch_size: config.batch_size,
+            events: config
+                .allowed_labels
+                .iter()
+                .map(|label| (label.clone(), 0))
+                .collect(),
+        })
+    }
+
+    /// Reports new events that should be included in the aggregated counts. If the number of
+    /// requests do not yet match the batch size `None` is returned. If the batch threshold
+    /// is reached the aggregated metrics for the batch are returned and the request count and
+    /// bucket counts are reset to 0. By default Laplacian noise is added to each of the aggregated
+    /// bucket counts that are returned. The metrics are returned as a tuple containing the batch
+    /// size and a vector of tuples cotaining the label and count for each bucket.
+    ///
+    /// If the `oak-unsafe` feature is enabled, or this function is called from a unit test, no
+    /// noise is added.
+    pub fn report_events(
+        &mut self,
+        events: HashSet<String>,
+    ) -> Option<(usize, Vec<(String, i64)>)> {
+        self.count += 1;
+        for (label, count) in self.events.iter_mut() {
+            if events.contains(label) {
+                *count += 1;
+            }
+        }
+
+        if self.count == self.batch_size {
+            let metrics = self
+                .events
+                .iter()
+                .map(|(label, count)| (label.to_string(), self.add_noise(*count)))
+                .collect();
+            self.reset();
+            Some((self.batch_size, metrics))
+        } else {
+            None
+        }
+    }
+
+    /// Adds no noise when the `oak-unsafe` feature is enabled or running as part of a unit test.
+    #[cfg(any(feature = "oak-unsafe", test))]
+    fn add_noise(&self, count: usize) -> i64 {
+        count as i64
+    }
+
+    /// Adds Laplacian noise by default.
+    #[cfg(all(not(feature = "oak-unsafe"), not(test)))]
+    fn add_noise(&self, count: usize) -> i64 {
+        self.add_laplace_noise(count)
+    }
+
+    /// Resets the request count and all the bucket counts to 0.
+    fn reset(&mut self) {
+        self.count = 0;
+        for (_, count) in self.events.iter_mut() {
+            *count = 0;
+        }
+    }
+
+    /// Adds Laplacian noise to a count. The Laplacian noise is sampled by sampling from a uniform
+    /// distribution and calculating the inverse of the Laplace cummulative distribution function on
+    /// the sampled value. Rounding of the noise is allowed as acceptable post-processing.
+    pub fn add_laplace_noise(&self, count: usize) -> i64 {
+        // Split the budget evenly over all of the labeled buckets.
+        let beta = self.events.len() as f64 / self.epsilon;
+        let p: f64 = thread_rng().sample(Open01);
+        count as i64 + Self::inverse_laplace(beta, p).round() as i64
```

Comment:
IIUC this means that if the count is zero, only half of the noise gets added because of the post-processing rounding, right? Is that fine?